File: src/main/java/com/example/demo/mapper/AccessLogMapper.java
package com.example.demo.mapper;

import com.example.demo.model.AccessLog;
import com.example.demo.model.IpLatestPath;
import org.apache.ibatis.annotations.Mapper;
import org.apache.ibatis.annotations.Param;

import java.util.List;

@Mapper
public interface AccessLogMapper {
    int insert(AccessLog record);

    List<IpLatestPath> selectLatestPathByIps(@Param("ips") List<String> ips);

    List<String> selectIpsWithMissingUserAgent();

    List<String> selectIpsByUserAgentPattern(
            @Param("pattern") String pattern,
            @Param("matchType") String matchType
    );
}

File: src/main/kotlin/com/example/demo/mapper/AccessLogMapper.kt
package com.example.demo.mapper

import com.example.demo.model.AccessLog
import com.example.demo.model.IpLatestPath
import org.apache.ibatis.annotations.Mapper
import org.apache.ibatis.annotations.Param

@Mapper
interface AccessLogMapper {
    fun insert(record: AccessLog): Int

    // 指定したIP一覧に対する最新アクセスのパス/UAを取得
    fun selectLatestPathByIps(@Param("ips") ips: List<String>): List<IpLatestPath>

    // user_agent が空（NULL/空白）のアクセスログを持つIP一覧（重複排除）
    fun selectIpsWithMissingUserAgent(): List<String>
}
